doc/app/user_function.rst

```diff
@@ -1,102 +1,167 @@
+    
+    .. note:: Non-composable apps cannot be added (or composed) with other apps.
+
+.. dropdown:: Handling not completed values
+
+    Cogent3 apps are designed to handle conditions under which data cannot be completely processed. For example, any errors during execution are stored in a :ref:`not completed <not_completed>` object (``NotCompleted``). All subsequent steps are skipped whenever an app returns a ``NotCompleted`` object.
+
+    These objects can also be used to stop the processing of a particular data record. For example, say you're writing an application that requires sequences to have some minimum length. If an input sequence is shorter than this, then you can create a :ref:`not completed <not_completed>` object and return it. This prevents any future processing, but the reasons for the failure (which you get to specify) can be saved for future reference.
+
+    You may want access to ``NotCompleted`` instances if, for example, you are developing an ``AppType.WRITER``. You can allow your code to "see" them with ``define_app(skip_not_completed=False)``.
 
-    from cogent3.app.typing import defined_types
+.. dropdown:: Supported cogent3 types
 
-    defined_types()
+    You can use the existing type hints if your function takes or returns ``cogent3`` types. To see these, use the ``defined_types()`` function.
 
-.. note:: You don't have to use cogent3 types, you can also use standard python types.
+    .. jupyter-execute::
 
-A simple example
-----------------
+        from cogent3.app.typing import defined_types
 
-Let's make an app that returns the elements of an alignment up to a specified index, with the index being a keyword argument. We now define a decorated function ``up_to`` and import the type hints and the decorator.
+        defined_types()
+
+    .. note:: You don't have to use cogent3 types. You can also use standard python types.
+
+Defining a cogent3 app from a function
+--------------------------------------
+
+We will write a function that takes a Cogent3 alignment and returns the first *n* positions where the user defines *n*.
 
 .. jupyter-execute::
 
     from cogent3.app.composable import define_app
     from cogent3.app.typing import AlignedSeqsType
 
     @define_app
-    def up_to(val: AlignedSeqsType, index=2) -> AlignedSeqsType:
-        return val[:index]
+    def n_positions(val: AlignedSeqsType, n=2) -> AlignedSeqsType:
+        return val[:n]
 
-We create an app instance for a specific value of ``index``
+The critical elements of a function being defined as an app are:
 
-.. jupyter-execute::
+1. The ``define_app`` decorator is used.
+2. Type hints are specified for the function's first argument and its return type.
 
-    first4 = up_to(index=4)
-    first4
+.. note: Currently, your function can only have one required argument.
```

Comment:
might be good to note that you can have multiple optional arguments in case people read this and decide it wont suit their use case